### ISSUE TITLE
[@realm/react] UserProvider blocks JS thread

### ISSUE THREAD

**Issue by thespacemanatee** [realm/realm-js#4715]
### How frequently does the bug occur?

All the time

### Description

When using `UserProvider`, after logging in the JS thread is constantly at <= 10fps. We have tried to minimise our code (rendering a `<View />` only on login) to pinpoint the cause of the issue but no print statements point to any infinite rerenders (caused by our code).

### Can you reproduce the bug?

Yes, always

### Reproduction Steps

```ts
const RootAppNavigator: FunctionComponent = () => {
  const publicRef = useRef<Realm>(null)
  const privateRef = useRef<Realm>(null)

  const realmApp = useApp()
  const realmUser = useUser()

  const mergedPublicConfig = useMemo(
    () => ({
      ...publicConfig,
      path: 'public.realm',
      sync: {
        ...publicConfig.sync,
        user: realmUser,
        partitionValue: PUBLIC_PARTITION,
        error: handlePublicRealmSyncError,
      } as Realm.SyncConfiguration,
    }),
    [realmUser, handlePublicRealmSyncError],
  )

  const mergedPrivateConfig = useMemo(
    () => ({
      ...privateConfig,
      path: `private.${realmUser?.id ?? ''}.realm`,
      sync: {
        ...privateConfig.sync,
        user: realmUser,
        partitionValue: realmUser?.id,
        error: handlePrivateRealmSyncError,
      } as Realm.SyncConfiguration,
    }),
    [realmUser, handlePrivateRealmSyncError],
  )

  return (
    <PublicRealmProvider realmRef={publicRef} {...mergedPublicConfig}>
      <PrivateRealmProvider realmRef={privateRef} {...mergedPrivateConfig}>
        <View /> // <-- literally rendering a <View /> only
      </PrivateRealmProvider>
    </PublicRealmProvider>
  )
}

const AppNavigator: FunctionComponent = () => {
  const { isLoading } = useAuth()

  return (
    <UserProvider
      fallback={
        <>
          <LoadingModal visible={isLoading} />
          <AuthStackNavigator />
        </>
      }
    >
      <RootAppNavigator />
    </UserProvider>
  )
}
```

### Version

11.0.0-rc.0

### What SDK flavour are you using?

Atlas App Services (auth, functions, etc.)

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

Android, iOS

### Build environment

```
"@realm/react": "^0.3.1",
"expo": "^45.0.6",
"react-native": "0.68.2",
"realm": "^11.0.0-rc.0"
```

### Cocoapods version

1.11.3

**Comment by takameyer** [comment]
@thespacemanatee thanks for creating this issue.  Did this start to happen after adding the `UserProvider`?

**Comment by thespacemanatee** [comment]
@takameyer Yeap, and it is reproducible all the time as the problem goes away whenever I remove the `UserProvider` and comes back after adding `UserProvider` and `useUser` back in.

**Comment by takameyer** [comment]
@thespacemanatee Good to know. I'll investigate this soon and get back to you.  Perhaps the event listener the `UserProvider` is using is firing too often.

**Comment by takameyer** [comment]
@thespacemanatee Can you pin down a specific component that perhaps is infinitely re-rendering?  If it's coming out of `@realm/react` is there a specific event this is coming from?

**Comment by thespacemanatee** [comment]
@takameyer thank you for investigating this issue! Will do some more digging on my end and see what I can do to create a simple repro, hopefully soon!

**Comment by kneth** [comment]
@thespacemanatee Any updates?

**Comment by thespacemanatee** [comment]
@kneth Sorry about that, I'm weeks away from graduation so didn't have much time to investigate. Let me get back to you guys this weekend.

**Comment by MrFuentastic** [comment]
@takameyer Hello, looks like I might have the same issue. The component I wrapped the UserProvider and 2 RealmProviders in is freezing with UserProvider. I tried removing the UserProvider and the screen works fine, the RealmProviders give a "user object must be a Realm user" error as expected. When I try the UserProvider alone (no RealmProviders) the screen also freezes.

I also have a similar Build Environment
```
"@realm/react": "^0.3.2",
"expo": "^45.0.0",
"react-native": "0.68.2",
"realm": "11.0.0-rc.0",
```

I'd share the repo but its my company's big legacy mobile app I'm trying to add some offline functionality to, if need I can try to reproduce in a new repo some time this week.
